File: 2026/09-September/12-Saturday/solution.py
from bisect import bisect_right
from functools import lru_cache

class Solution:
    def maximumWeight(self, intervals):
        n = len(intervals)

        # Store: (start, end, weight, original_index)
        arr = [
            (s, e, w, i)
            for i, (s, e, w) in enumerate(intervals)
        ]

        # Sort by start, then end, then original index
        arr.sort()

        starts = [x[0] for x in arr]

        # next[i] = first interval that starts after arr[i]'s end
        nxt = [0] * n

        for i in range(n):
            nxt[i] = bisect_right(starts, arr[i][1])

        @lru_cache(None)
        def dp(i, k):
            if i >= n or k == 0:
                return (0, ())

            # Option 1: skip current interval
            score1, indices1 = dp(i + 1, k)

            # Option 2: take current interval
            score2, indices2 = dp(nxt[i], k - 1)
            score2 += arr[i][2]

            candidate = tuple(sorted(indices2 + (arr[i][3],)))

            if score2 > score1:
                return score2, candidate

            if score2 < score1:
                return score1, indices1

            # Same score -> lexicographically smaller indices
            return score1, min(indices1, candidate)

        return list(dp(0, 4)[1])

File: 2026/09-September/12-Saturday/solution.cpp
class Solution {
public:
    vector<int> maximumWeight(vector<vector<int>>& intervals) {
        int n = intervals.size();

        // {start, end, weight, original index}
        vector<array<long long, 4>> a;

        for (int i = 0; i < n; i++) {
            a.push_back({
                intervals[i][0],
                intervals[i][1],
                intervals[i][2],
                i
            });
        }

        // Sort by start time
        sort(a.begin(), a.end());

        vector<long long> starts(n);
        for (int i = 0; i < n; i++) {
            starts[i] = a[i][0];
        }

        // next[i] = first interval with start > a[i].end
        vector<int> nxt(n);

        for (int i = 0; i < n; i++) {
            nxt[i] = upper_bound(
                starts.begin(),
                starts.end(),
                a[i][1]
            ) - starts.begin();
        }

        // dp[i][k] = best result from i onward
        // with at most k intervals.
        //
        // Store {score, indices}
        struct State {
            long long score;
            vector<int> indices;
        };

        vector<vector<State>> dp(n + 1, vector<State>(5));
        vector<vector<bool>> vis(n + 1, vector<bool>(5, false));

        function<State(int, int)> solve = [&](int i, int k) -> State {
            if (i >= n || k == 0)
                return {0, {}};

            if (vis[i][k])
                return dp[i][k];

            vis[i][k] = true;

            // Option 1: skip current interval
            State skip = solve(i + 1, k);

            // Option 2: take current interval
            State take = solve(nxt[i], k - 1);
            take.score += a[i][2];
            take.indices.push_back((int)a[i][3]);

            // Keep indices sorted for lexicographical comparison
            sort(take.indices.begin(), take.indices.end());

            State ans;

            if (take.score > skip.score) {
                ans = take;
            } 
            else if (take.score < skip.score) {
                ans = skip;
            } 
            else {
                // Same score -> lexicographically smaller indices
                if (take.indices < skip.indices)
                    ans = take;
                else
                    ans = skip;
            }

            return dp[i][k] = ans;
        };

        return solve(0, 4).indices;
    }
};
